crates/network/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+                tracing::warn!("error writing data to stream: {e}")
+            }
+        }
+    });
+
+    let connection = RemoteConnection {
+        remote_addr,
+        recv: incoming_receiver,
+        send: outgoing_sender,
+        encoder: PacketEncoder::default(),
+        recv_task,
+        send_task,
+        state: HandshakeState::Handshaking,
+    };
+
+    let _ = shared.0.new_connections_send.send_async(connection).await;
+}
+
+#[allow(clippy::type_complexity)]
+fn run_packet_event_loop(
+    world: &mut World,
+    state: &mut SystemState<(
+        Query<(Entity, &mut RemoteConnection)>,
+        EventWriter<PacketReceivedEvent>,
+        EventWriter<DisconnectEvent>,
+        Commands,
+    )>,
+) {
+    let (mut connections, mut packet_events, mut disconnect_events, mut commands) =
+        state.get_mut(world);
+
+    for (entity, mut connection) in &mut connections {
+        let result = connection.try_recv();
+
+        match result {
+            Ok(frame) => {
+                packet_events.send(PacketReceivedEvent {
+                    connection: entity,
+                    frame,
+                });
+            }
+            Err(TryRecvError::Empty) => {}
+            Err(TryRecvError::Disconnected) => {
+                tracing::info!("Client disconnected");
+                disconnect_events.send(DisconnectEvent { connection: entity });
+                commands.entity(entity).remove::<RemoteConnection>();
+            }
+        }
+    }
+
+    state.apply(world);
+}
+
+fn connection_event_handler(
+    mut packets: EventReader<PacketReceivedEvent>,
+    mut handshake_events: EventWriter<HandshakeEvent>,
+    mut client_status_request_events: EventWriter<ClientStatusRequestEvent>,
+    mut connections: Query<&mut RemoteConnection>,
+) {
+    for packet in packets.iter() {
+        let payload = &packet.frame.payload;
+        if let Ok(mut connection) = connections.get_mut(packet.connection) {
+            match &connection.state {
+                HandshakeState::Handshaking => {
+                    match ClientHandshakePacket::decode(&mut std::io::Cursor::new(payload)) {
+                        Ok(p) => match p {
+                            ClientHandshakePacket::Handshake(p) => {
+                                handshake_events.send(HandshakeEvent {
+                                    connection: packet.connection,
+                                    protocol_version: p.protocol_version,
+                                    server_address: p.server_address,
+                                    server_port: p.server_port,
+                                    next_state: p.next_state,
+                                });
+                                connection.state = p.next_state;
+                            }
+                        },
+                        Err(e) => {
+                            tracing::error!("error: {:?}", e)
```

Comment:
Why don't we report the error to the caller via `Result` making it handle the problem?

Reply:
The caller is Bevy's system scheduler

Reply:
Hmm, should not we at least receive a handle to cancel client communication in case of error? I am not sure, but as for now it seems unpredictable to me what happens when an error ocurrs.